### ISSUE TITLE
Cursor-Based Pagination

### ISSUE THREAD

**Issue by richardklint** [phpclassic/php-shopify#101]
With the latest API release, certain pagination calls are not working. See https://help.shopify.com/en/api/guides/paginated-rest-results

**Comment by tareqtms** [comment]
@richardklint 
We don't have any specific functions to handle paginated results. What update do you want from us?

**Comment by richardklint** [comment]
This is just to track progress towards getting the new pagination working.

**Comment by KiloooNL** [comment]
> 
> 
> This is just to track progress towards getting the new pagination working.

Pagination is quite easy, this works for me, it may help you:

`                // Shopify paginiation
                $productsCount = $shopify->Product->count();
                echo "<br>Total Shopify products in Shopify: " . $productsCount;

                $page = 1;   // First page
                $limit = 25; // Limit of products per page
                $getShopifyProducts = array(); // Create an array for Shopify products
                
                // Work out how many pages there are
                $totalPages = ceil($productsCount / $limit);
                while ($page <= $totalPages) {
                    echo '<br><b>Current page: ' . $page . ' / ' . $totalPages . '</b>';
                    $params = array("limit" => $limit, "page" => $page);
                    array_push($getShopifyProducts, $shopify->Product->get($params));

                    // Go to the next page.
                    $page++;
                }`

**Comment by zealnagar** [comment]
@KiloooNL you are right but after the Curson based pagination in 2019-07 and above they provide `next` and `previous` link in header when you pass limit query string parameter. So we need support to access CurlResponse class getHeader method statically. So we can iterate the pages and get relevant info based on the individual use case.

**Comment by shawnhind** [comment]
You can also use since_id based pagination which they still say is the most efficient (moreso than cursor-based) and has been available in the API for quite a long time.

This would be a nice convenience thing to have though.

> Using a since_id is the fastest solution and should be the first choice for pagination when order doesn’t matter.

Ref:
https://www.shopify.com/partners/blog/relative-pagination

**Comment by tareqtms** [comment]
@zealnagar 
Ok, will have a look before the next release.

**Comment by mailramk** [comment]
Hi,
When can we expect this enhancement implemented as the support for pagination is stopping from April 2020. This is the screenshot from the mail received from Shopify about this.

![2019-10-22_08h51_46](https://user-images.githubusercontent.com/21346566/67286920-2db48880-f4a8-11e9-9032-b2be6c03029a.png)

Thanks
Ram

**Comment by rickcow** [comment]
Hi, this issue should be in top priority. If page filter is removed on April 1, 2020, this SDK will be useless.  I am looking forward to hear there is a support for this soon. 

Thanks
Rick

**Comment by shawnhind** [comment]
@rickcow take a look at my comment.  since_id is supported in this API and has nothing to do with page filtering.  You can use this in this SDK in its current state very easily.

**Comment by shawnhind** [comment]
@Coennie oh okay.  Thanks, wasn't aware that this wasn't available for all endpoints, yeah in that case this would be necessary then.  I only use the Page, Blog, Product, and ProductCollection endpoints which all have it so I just made an assumption that they all did.

**Comment by alexgorney** [comment]
> @rickcow take a look at my comment above in this thread. since_id pagination is still supported in this new API and has nothing to do with page filtering being removed. You can use this in this SDK in its current state very easily and isn't going away and is actually even more efficient than cursor based pagination.
> 
> Example to retrieve all Products:
> 
> ```
> $lastId = 0;
> $products = [];
> 
> do {
>     $results = $client->Product->get(['limit' => 250, 'since_id' => $lastId]);
>     $products = array_merge($products, $results);
>     $lastId = end($results)['id'];
> } while (sizeof($results) > 0);
> ```

Be careful with this as I don't believe the products are necessarily returned in `id` ASC order.
Might be a good idea to use 'order' => 'created_at ASC' as a param as well.

**Comment by shamsicodebase** [comment]
This is just to track progress towards getting the new pagination working. I also emailed you. Please reply. 
Thank You

**Comment by tareqtms** [comment]
Hello Everyone,
I am extremely sorry to keep you waiting. I will keep it at the top priority for the next couple of weeks and hopefully come up with a solution. 
Thanks

**Comment by ranjitatwal30** [comment]
Awaiting for an update on this one...

**Comment by zamliyconsulting** [comment]
> Hello Everyone,
> I am extremely sorry to keep you waiting. I will keep it at the top priority for the next couple of weeks and hopefully come up with a solution.
> Thanks
Is this helpful?
https://community.shopify.com/c/Shopify-APIs-SDKs/How-to-parse-Link-data-from-header-data-in-PHP/m-p/570407#M38050

**Comment by tareqtms** [comment]
Hi everyone,

I have created a new release today (v1.1.6) with an available pull request for the new pagination system. I reviewed the codes and fixed some coding issues, but I couldn't test it 100%. At the final build the next page could be retrieved by the following example code:

```php
$productResource =  $shopify->getApi()->Product();
$products = $productResource->get(['limit'=>50]);
$nextPageProducts = $productResource->get($productResource->getNextPageParams());
/* in short: you need to reuse the resource to get the params before making another resource */**
```

**Comment by zealnagar** [comment]
@tareqtms I have tested the change. I found one bug for getting the previous link and I have fixed.
For that Pull Request has been created. You can review and merge it if proper.
For getting paginated data following code worked for me.

```
$productResource =  $shopify->Product();
$products = $productResource->get(['limit'=>50]);
$nextPageProducts = $productResource->get($productResource->getNextPageParams());
```
`getApi()` method gave me error. So i remoed ant it worked.

**Comment by tareqtms** [comment]
@zealnagar 
Thanks, I have merged the pull request and created a new build v1.1.7.
I also removed the `getApi()` function from my comment, that was wrong and was giving me the same problem, but forgot to remove that part when copying the codes from another comment.

**Comment by zamliyconsulting** [comment]
> @tareqtms I have tested the change. I found one bug for getting the previous link and I have fixed.
> For that Pull Request has been created. You can review and merge it if proper.
> For getting paginated data following code worked for me.
> 
> ```
> $productResource = $shopify->Product();
> $products        = $productResource->get(['limit' => 250]);
> $next_page       = $productResource->getNextPageParams();
> while ($next_page) {
>     $new_products = $productResource->get($productResource->getNextPageParams());
>     $products     = array_merge($products, $new_products);
>     $next_page    = $productResource->getNextPageParams();
> }
> dd(json_encode($products));
> ```
> 
This does not work for me and I have no idea what I am doing wrong.  It keeps on giving me 250 products instead of the 2317 products i have.

$temp = getProductsNEW();
echo count($temp);

function getProductsNEW() {
	global $shopify;
	$productResource = $shopify->Product();
	$products        = $productResource->get(['limit' => 250]);
	$next_page       = $productResource->getNextPageParams();
	while ($next_page) {
		$new_products = $productResource->get($productResource->getNextPageParams());
		$products     = array_merge($products, $new_products);
		$next_page    = $productResource->getNextPageParams();
	}
	return $products;
}

**Comment by zealnagar** [comment]
@zamliyconsulting I have copied your code and executed it on my system. It is working.
In the output console, 1st execution is with `limit 1` filter, and in 2nd execution is with `limit 250`.
In both the case, I got count as 4 since I have 4 products in my store.

Don't know why you are getting only 250. Please make sure the following requirement meets.

1. Update the package to latest version i.e v1.1.7
2. While setting SDK config, you have passed API version >= 2019-07.

![Screenshot 2020-02-18 at 10 48 33 AM](https://user-images.githubusercontent.com/29673644/74706438-fdee7500-523c-11ea-860c-3826e2a31b8c.png)
![Screenshot 2020-02-18 at 10 47 11 AM](https://user-images.githubusercontent.com/29673644/74706443-0181fc00-523d-11ea-9575-39ef94934e63.png)

**Comment by zamliyconsulting** [comment]
> @zamliyconsulting I have copied your code and executed it on my system. It is working.
> In the output console, 1st execution is with `limit 1` filter, and in 2nd execution is with `limit 250`.
> In both the case, I got count as 4 since I have 4 products in my store.
> 
> Don't know why you are getting only 250. Please make sure the following requirement meets.
> 
> 1. Update the package to latest version i.e v1.1.7
> 2. While setting SDK config, you have passed API version >= 2019-07.
> 
> ![Screenshot 2020-02-18 at 10 48 33 AM](https://user-images.githubusercontent.com/29673644/74706438-fdee7500-523c-11ea-860c-3826e2a31b8c.png)
> ![Screenshot 2020-02-18 at 10 47 11 AM](https://user-images.githubusercontent.com/29673644/74706443-0181fc00-523d-11ea-9575-39ef94934e63.png)

I downloaded the code yesterday, so I assume it is the latest version.  API version is already passed in the ShopifySDK.php file (though line 313 was left as 2019-04), so I do not include it in my config array. 

Well, since the purpose of this code is to test with more than 250 products, why are you testing with only 4 products?  You don't need any of this for 4 products.

**Comment by tareqtms** [comment]
@zamliyconsulting 
I think you are facing the issue because of the default version is '2019-04'. You need '2019-07' to make the cursor-based pagination work. 
I should have made the change to the default version actually. Will do in the next build.

**Comment by tareqtms** [comment]
Just created a new release with the ApiVersion fix.

**Comment by zamliyconsulting** [comment]
> Just created a new release with the ApiVersion fix.

Thank you but I had already fixed it.  Nonetheless, I downloaded it and tested with same results.  Has anybody actually tested this with more than 250 products in their system?

**Comment by tareqtms** [comment]
@zamliyconsulting 
Are you confirming that it's not working for you after updating the `ApiVersion` too?

**Comment by zamliyconsulting** [comment]
Yes, I just downloaded your latest version 5 minutes ago.  Same result.

**Comment by zamliyconsulting** [comment]
> @zamliyconsulting
> Are you confirming that it's not working for you after updating the `ApiVersion` too?

Yes, I just downloaded your latest version 5 minutes ago. Same result.  I also tried including the api version in the config array, same result.  $next_page array before the loop is coming empty even though I have 2317 products.

**Comment by zealnagar** [comment]
@zamliyconsulting  I have tested with more than 250 products please see the attached screenshot.

![Screenshot 2020-02-19 at 4 51 36 PM](https://user-images.githubusercontent.com/29673644/74829956-24450b00-5338-11ea-8332-e67b76195d44.png)

**Comment by zamliyconsulting** [comment]
Are you using your own code or did you download the latest version here?  Your pull code and what is in the ShopifyResource.php are not the same.  I am not sophisticated enough to understand the difference, but I can see that they are different.

**Comment by zamliyconsulting** [comment]
> @zamliyconsulting I have tested with more than 250 products please see the attached screenshot.
> It is working absolutely fine.
> 
> ![Screenshot 2020-02-19 at 4 51 36 PM](https://user-images.githubusercontent.com/29673644/74829956-24450b00-5338-11ea-8332-e67b76195d44.png)

Are you using your own code or did you download the latest version here? Your pull code and what is in the ShopifyResource.php are not the same. I am not sophisticated enough to understand the difference, but I can see that they are different.

**Comment by zamliyconsulting** [comment]
> @zamliyconsulting I have tested with more than 250 products please see the attached screenshot.
> It is working absolutely fine.
> 
> ![Screenshot 2020-02-19 at 4 51 36 PM](https://user-images.githubusercontent.com/29673644/74829956-24450b00-5338-11ea-8332-e67b76195d44.png)

Are you using your own code or did you download the latest version here? Your pull code and what is in the ShopifyResource.php are not the same. I am not sophisticated enough to understand the difference, but I can see that they are different.

**Comment by zealnagar** [comment]
@zamliyconsulting I have updated to the latest version of the package. see my composer.lock

![Screenshot 2020-02-19 at 5 45 52 PM](https://user-images.githubusercontent.com/29673644/74833575-c1f00880-533f-11ea-84b2-d9627af4b031.png)

**Comment by zamliyconsulting** [comment]
> @zamliyconsulting I have updated to the latest version of the package. see my composer.lock
> 
Why does mine look different if we are downloading the same package?
![image](https://user-images.githubusercontent.com/33092049/74835308-08674800-52eb-11ea-9965-1dea2bd666e8.png)

**Comment by zamliyconsulting** [comment]
> @zamliyconsulting I have updated to the latest version of the package. see my composer.lock
> 
Why does mine look different is we downloaded the same package?
![image](https://user-images.githubusercontent.com/33092049/74836156-d2c35e80-52ec-11ea-925f-a6069eb8f979.png)

**Comment by zealnagar** [comment]
@zamliyconsulting let's connect on Google meet and I will help you with this. Please send me google meet link 
`zealnagar.nagar@gmail.com`

**Comment by zamliyconsulting** [comment]
> @zamliyconsulting let's connect on Google meet and I will help you with this. Please send me google meet link
> `zealnagar.nagar@gmail.com`

I sent you an email but can you please confirm that these are the exact same files you are working with (i changed the extensions as txt)  I simply do not see how this can work for you but not for me.

[ShopifyResource.txt](https://github.com/phpclassic/php-shopify/files/4227614/ShopifyResource.txt)
[CurlResponse.txt](https://github.com/phpclassic/php-shopify/files/4227615/CurlResponse.txt)
[CurlRequest.txt](https://github.com/phpclassic/php-shopify/files/4227616/CurlRequest.txt)

**Comment by zamliyconsulting** [comment]
> @zamliyconsulting let's connect on Google meet and I will help you with this. Please send me google meet link
> `zealnagar.nagar@gmail.com`

I checked to see wehat the code does in the following function and realized that my query does not fall under either of the if clauses, hence giving me a blank $next_page array.
**public function getLink($responseHeaders, $type='next')**

**Comment by zamliyconsulting** [comment]
> @zamliyconsulting
> Are you confirming that it's not working for you after updating the `ApiVersion` too?
@zealnagar @tareqtms 

OK, i finally found the problem.  I still don't how this was working for you @zealnagar?

@tareqtms - You will need the fix the following in shopifyResource.php:
On lines 558 and 559, change x-shopify-api-version to X-Shopify-API-Version
On lines 569,570,571, change ['link'] to ['Link']

NOW, it is working!!!!

**Comment by zamliyconsulting** [comment]
How does this work with Order or InventoryLevel, especially when you need to have parameters other than 'limit'?  I could not make it work.

**Comment by zealnagar** [comment]
> How does this work with Order or InventoryLevel, especially when you need to have parameters other than 'limit'? I could not make it work.

Here is the solution for parameter other than limit

```
$filters = [
    'limit'              => 250,
    'fulfillment_status' => 'unfulfilled',
];
$order_resource = $shopify->Order();
$orders         = $order_resource->get($filters);
$next_page      = $order_resource->getNextPageParams();
while ($next_page) {
    $filters          = array_merge($filters, $next_page);
    $next_page_orders = $order_resource->get($filters);
    $orders           = array_merge($orders, $next_page_orders);
    $next_page        = $order_resource->getNextPageParams();
}
```

now in the $filter you can pass other parameters to like I have passed.

**Comment by arosenhagen** [comment]
> > @zamliyconsulting
> > Are you confirming that it's not working for you after updating the `ApiVersion` too?
> > @zealnagar @tareqtms
> 
> OK, i finally found the problem. I still don't know how this was working for you @zealnagar?
> 
> @tareqtms - You will need the fix the following in shopifyResource.php:
> On lines 558 and 559, change x-shopify-api-version to X-Shopify-API-Version
> On lines 563,564,565, change ['link'] to ['Link']
> 
> NOW, it is working!!!!

This works for me too. We had the problem, that executing an API logic from a local php file went well but running it via a serverless function in AWS lambda only returned those 250 product limit.

**Comment by zamliyconsulting** [comment]
>     $filters          = array_merge($filters, $next_page);

While your coding is correct, this does not for many reasons:
1) Orders cursor based pagination requires API Version 2019-10, 2019-07 does not work.
2) Orders cursor based pagination allows the passing of only 3 filter values: page, limit, and field names.

I guess we have use the since_id option to find specific time range orders.  What a mess!

**Comment by zamliyconsulting** [comment]
For InventoryLevel, it says I can't pass location_ids when page is present, yet it is a required variable.  I must be doing something wrong!

Fatal error: Uncaught PHPShopify\Exception\ApiException: location_ids - location_ids cannot be passed when page_info is present.

**Comment by zamliyconsulting** [comment]
> > How does this work with Order or InventoryLevel, especially when you need to have parameters other than 'limit'? I could not make it work.
> 
> Here is the solution for parameter other than limit
> 
> ```
> $filters = [
>     'limit'              => 250,
>     'fulfillment_status' => 'unfulfilled',
> ];
> $order_resource = $shopify->Order();
> $orders         = $order_resource->get($filters);
> $next_page      = $order_resource->getNextPageParams();
> while ($next_page) {
>     $filters          = array_merge($filters, $next_page);
>     $next_page_orders = $order_resource->get($filters);
>     $orders           = array_merge($orders, $next_page_orders);
>     $next_page        = $order_resource->getNextPageParams();
> }
> ```
> 
> now in the $filter you can pass other parameters to like I have passed.

This is incorrect.  You can pass the filters ONLY the first time.  Here is the corrected version. It works in this format.

$filters = [
    'limit'              => 250,
    'fulfillment_status' => 'unfulfilled',
];
$order_resource = $shopify->Order();
$orders         = $order_resource->get($filters);
$next_page      = $order_resource->getNextPageParams();
while ($next_page) {
    $next_page_orders = $order_resource->get($order_resource->getNextPageParams());
    $orders           = array_merge($orders, $next_page_orders);
    $next_page        = $order_resource->getNextPageParams();
}

**Comment by raviMukti** [comment]
> 
> 
> > > How does this work with Order or InventoryLevel, especially when you need to have parameters other than 'limit'? I could not make it work.
> > 
> > 
> > Here is the solution for parameter other than limit
> > ```
> > $filters = [
> >     'limit'              => 250,
> >     'fulfillment_status' => 'unfulfilled',
> > ];
> > $order_resource = $shopify->Order();
> > $orders         = $order_resource->get($filters);
> > $next_page      = $order_resource->getNextPageParams();
> > while ($next_page) {
> >     $filters          = array_merge($filters, $next_page);
> >     $next_page_orders = $order_resource->get($filters);
> >     $orders           = array_merge($orders, $next_page_orders);
> >     $next_page        = $order_resource->getNextPageParams();
> > }
> > ```
> > 
> > 
> > now in the $filter you can pass other parameters to like I have passed.
> 
> This is incorrect. You can pass the filters ONLY the first time. Here is the corrected version. It works in this format both for InventoryLevel and Order. BTW, you need API Version 2019-10 for these. They do not work with 2019-07.
> 
> ```
> $filters = [
>     'limit'              => 250,
>     'fulfillment_status' => 'unfulfilled',
> ];
> $order_resource = $shopify->Order();
> $orders         = $order_resource->get($filters);
> $next_page      = $order_resource->getNextPageParams();
> while ($next_page) {
>     $next_page_orders = $order_resource->get($order_resource->getNextPageParams());
>     $orders           = array_merge($orders, $next_page_orders);
>     $next_page        = $order_resource->getNextPageParams();
> }
> ```

@zamliyconsulting hi zamliy, I have tried the code above, but this gives me a very long time to get a response, does this repeat up to all the orders in the order history? I provide a filter for orders that are paid and also unfulfilled. I have checked the total order on the dashboard, there are only 500 more orders, but this takes approximately 4 minutes

```
$shopify = new ShopifySDK($config);
        $filters = [
            'limit' => 250,
            'financial_status' => 'paid',
            'fulfillment_status' => 'unfulfilled'
        ];
        $orders_resource = $shopify->Order();
        $orders = $orders_resource->get($filters);
        $next_page = $orders_resource->getNextPageParams();
        while ($next_page) {
            $next_page_orders = $orders_resource->get($orders_resource->getNextPageParams());
            $orders = array_merge($orders, $next_page_orders);
            $next_page = $orders_resource->getNextPageParams();
        }
        return $orders;
```

**Comment by zamliyconsulting** [comment]
> @zamliyconsulting hi zamliy, I have tried the code above, but this gives me a very long time to get a response, does this repeat up to all the orders in the order history? I provide a filter for orders that are paid and also unfulfilled. I have checked the total order on the dashboard, there are only 500 more orders, but this takes approximately 4 minutes
> 
> ```
> $shopify = new ShopifySDK($config);
>         $filters = [
>             'limit' => 250,
>             'financial_status' => 'paid',
>             'fulfillment_status' => 'unfulfilled'
>         ];
>         $orders_resource = $shopify->Order();
>         $orders = $orders_resource->get($filters);
>         $next_page = $orders_resource->getNextPageParams();
>         while ($next_page) {
>             $next_page_orders = $orders_resource->get($orders_resource->getNextPageParams());
>             $orders = array_merge($orders, $next_page_orders);
>             $next_page = $orders_resource->getNextPageParams();
>         }
>         return $orders;
> ```

Nobody accused Shopify of being fast :-)    You can try adding 'created_at_min' and/or 'created_at_max' to the filters to see if gets faster.

**Comment by raviMukti** [comment]
> 
> 
> > @zamliyconsulting hi zamliy, I have tried the code above, but this gives me a very long time to get a response, does this repeat up to all the orders in the order history? I provide a filter for orders that are paid and also unfulfilled. I have checked the total order on the dashboard, there are only 500 more orders, but this takes approximately 4 minutes
> > ```
> > $shopify = new ShopifySDK($config);
> >         $filters = [
> >             'limit' => 250,
> >             'financial_status' => 'paid',
> >             'fulfillment_status' => 'unfulfilled'
> >         ];
> >         $orders_resource = $shopify->Order();
> >         $orders = $orders_resource->get($filters);
> >         $next_page = $orders_resource->getNextPageParams();
> >         while ($next_page) {
> >             $next_page_orders = $orders_resource->get($orders_resource->getNextPageParams());
> >             $orders = array_merge($orders, $next_page_orders);
> >             $next_page = $orders_resource->getNextPageParams();
> >         }
> >         return $orders;
> > ```
> 
> Nobody accused Shopify of being fast :-) You can try adding 'created_at_min' and/or 'created_at_max' to the filters to see if gets faster.

haha okay if that's the case I'll try your advice first, thanks in advance

**Comment by raviMukti** [comment]
> 
> 
> > @zamliyconsulting hi zamliy, I have tried the code above, but this gives me a very long time to get a response, does this repeat up to all the orders in the order history? I provide a filter for orders that are paid and also unfulfilled. I have checked the total order on the dashboard, there are only 500 more orders, but this takes approximately 4 minutes
> > ```
> > $shopify = new ShopifySDK($config);
> >         $filters = [
> >             'limit' => 250,
> >             'financial_status' => 'paid',
> >             'fulfillment_status' => 'unfulfilled'
> >         ];
> >         $orders_resource = $shopify->Order();
> >         $orders = $orders_resource->get($filters);
> >         $next_page = $orders_resource->getNextPageParams();
> >         while ($next_page) {
> >             $next_page_orders = $orders_resource->get($orders_resource->getNextPageParams());
> >             $orders = array_merge($orders, $next_page_orders);
> >             $next_page = $orders_resource->getNextPageParams();
> >         }
> >         return $orders;
> > ```
> 
> Nobody accused Shopify of being fast :-) You can try adding 'created_at_min' and/or 'created_at_max' to the filters to see if gets faster.

@zamliyconsulting hi zamliy, I get an error like this PHP Fatal error:  Allowed memory size of 1073741824 bytes exhausted (tried to allocate 1679360 bytes) it's because the infinite loop isn't?

i have updated my code to something like this :
```
$shopify = new ShopifySDK($config);
        $filters = [
            'limit' => 250,
            'financial_status' => 'paid',
            'fulfillment_status' => 'unfulfilled',
            'created_at_min ' => '2020-04-09',
            'created_at_max' => '2020-04-10'
        ];
        $orders_resource = $shopify->Order();
        $orders = $orders_resource->get($filters);
        $next_page = $orders_resource->getNextPageParams();
        while ($next_page) {
            $next_page_orders = $orders_resource->get($orders_resource->getNextPageParams());
            $orders = array_merge($orders, $next_page_orders);
            $next_page = $orders_resource->getNextPageParams();
        }
        return $orders;
```

**Comment by zamliyconsulting** [comment]
> @zamliyconsulting hi zamliy, I get an error like this - Allowed memory size of 1073741824 bytes exhausted (tried to allocate 1679360 bytes) in D:\Project\Web\marketplace-slim\vendor\phpclassic\php-shopify\lib\CurlResponse.php on line 22 it's because the infinite loop isn't?
> 
> i have updated my code to something like this :
> 
> ```
> $shopify = new ShopifySDK($config);
>         $filters = [
>             'limit' => 250,
>             'financial_status' => 'paid',
>             'fulfillment_status' => 'unfulfilled',
>             'created_at_min ' => '2020-04-09',
>             'created_at_max' => '2020-04-10'
>         ];
>         $orders_resource = $shopify->Order();
>         $orders = $orders_resource->get($filters);
>         $next_page = $orders_resource->getNextPageParams();
>         while ($next_page) {
>             $next_page_orders = $orders_resource->get($orders_resource->getNextPageParams());
>             $orders = array_merge($orders, $next_page_orders);
>             $next_page = $orders_resource->getNextPageParams();
>         }
>         return $orders;
> ```

You need the Y-m-dTH:i:s-P date format for those 2 date.  Also, try ini_set('memory_limit', '-1'); on top of your page.

**Comment by cbepxpa3ym** [comment]
This is crazy how you guys use pagination with "while". Here is the correct way to do so with "do-while":

```php
$productResource = $shopify->Product();
$pageParams = [
    'limit' => 100,
    'product_type' => 'some type',
    'fields' => 'id,variants'
];
do {
    $productsList = $productResource->get($pageParams);
    // place your logic here
    $pageParams = $productResource->getNextPageParams();
} while($pageParams);
```

**Comment by hurta2yaisel** [comment]
I'm trying to get the orders with pagination using this code and I can't get the next page info. I can only get the last 15 orders. I'm using the version 1.1.14 of this library (default ApiVersion 2020-01). What I'm doing wrong?

`
$shopify = ShopifySDK::config($config);
$orderParams = array(
    "financial_status" => "paid",
    "limit" => 50
);

$orderResource = $shopify->Order();
$pending_orders = array();

do {
    $json_data = $orderResource->get($orderParams);
    $pending_orders = array_merge($pending_orders, $json_data);
    $orderParams = $orderResource->getNextPageParams();

    file_put_contents('recovery.json', json_encode($pending_orders));
    //var_dump($json_data);
    var_dump($orderParams);
    var_dump(count($pending_orders));
    var_dump($orderResource->getNextLink());
} while ($orderParams);
`

**Comment by zamliyconsulting** [comment]
> I'm trying to get the orders with pagination using this code and I can't get the next page info. I can only get the last 15 orders. I'm using the version 1.1.14 of this library (default ApiVersion 2020-01). What I'm doing wrong?
> 
> ```
> $shopify = ShopifySDK::config($config);
> $orderParams = array(
>     "financial_status" => "paid",
>     "limit" => 50
> );
> $orderResource = $shopify->Order();
> $pending_orders = array();
> do {
>     $json_data = $orderResource->get($orderParams);
>     $pending_orders = array_merge($pending_orders, $json_data);
>     $orderParams = $orderResource->getNextPageParams();
> 
>     file_put_contents('recovery.json', json_encode($pending_orders));
>     //var_dump($json_data);
>     var_dump($orderParams);
>     var_dump(count($pending_orders));
>     var_dump($orderResource->getNextLink());
> } while ($orderParams);
> ```
I will give you what is working.  You can figure out the rest.

	$filters = [
		"status" => "any",
		"limit" => 250
	];

	$orderResource = $shopify->Order();
	$orders = $orderResource->get($filters);
	$next_page = $orderResource->getNextPageParams();
	while ($next_page) {
		$new_orders = $orderResource->get($orderResource->getNextPageParams());
		$orders = array_merge($orders, $new_orders);
		$next_page = $orderResource->getNextPageParams();
	}
	return $orders;

**Comment by cbepxpa3ym** [comment]
> I will give you what is working.

The code he uses works too. And that code (the loop) is professional comparing to what you have provided. If you do not understand the difference then you have to change your specialty, I think. lol

**Comment by cbepxpa3ym** [comment]
> What I'm doing wrong?

Do you work on the Private app? If not then you have to [request access to all orders](https://shopify.dev/tutorials/authenticate-a-public-app-with-oauth#orders-permissions). Also are you sure you have more than 15 paid orders?

**Comment by hurta2yaisel** [comment]
> > What I'm doing wrong?
> 
> Do you work on the Private app? If not then you have to [request access to all orders](https://shopify.dev/tutorials/authenticate-a-public-app-with-oauth#orders-permissions). Also are you sure you have more than 15 paid orders?

It is a public application and yep, I have 54 paid orders. Let me try requesting access to all orders.

**Comment by hurta2yaisel** [comment]
> > What I'm doing wrong?
> 
> Do you work on the Private app? If not then you have to [request access to all orders](https://shopify.dev/tutorials/authenticate-a-public-app-with-oauth#orders-permissions). Also are you sure you have more than 15 paid orders?

I found what I was missing. I had to include the filter by status, otherwise it only brought the unfulfilled orders.

```
$shopify = ShopifySDK::config($config);
$orderParams = array(
    **"status" => "any",**
    "financial_status" => "paid",
    "limit" => 50
);
$orderResource = $shopify->Order();
$pending_orders = array();
do {
    $json_data = $orderResource->get($orderParams);
    $pending_orders = array_merge($pending_orders, $json_data);
    $orderParams = $orderResource->getNextPageParams();

    file_put_contents('recovery.json', json_encode($pending_orders));
    var_dump($orderParams);
} while ($orderParams);
var_dump($pending_orders);
var_dump(count($pending_orders));
```

**Comment by joeyc002** [comment]
This certainly doesn't work for InventoryLevel. 

`//Get inventory levels
$page_limit = 5;
$filters = [
	'location_ids'		 => $this->store_settings['shopify_location_id'],
	'limit'              => $page_limit
];
$inventory_resource = $this->shopify->InventoryLevel();
$all_inventory_levels  = $inventory_resource->get($filters);
$next_page  	    = $inventory_resource->getNextPageParams();

while ($next_page) {
	$filters          = array_merge($filters, $next_page);
	$next_page_inventory_level = $inventory_resource->get(array('limit' => $page_limit));
	$all_inventory_levels = array_merge($orders, $next_page_inventory_level);
	$next_page        = $inventory_resource->getNextPageParams();
}

print_r($all_inventory_levels);
`

If it has to go on to the next page, it throws a HTTP 422 error.

**Comment by lexthor** [comment]
> Hi everyone,
> 
> I have created a new release today (v1.1.6) with an available pull request for the new pagination system. I reviewed the codes and fixed some coding issues, but I couldn't test it 100%. At the final build the next page could be retrieved by the following example code:
> 
> ```
> $productResource =  $shopify->Product();
> $products = $productResource->get(['limit'=>50]);
> $nextPageProducts = $productResource->get($productResource->getNextPageParams());
> /* in short: you need to reuse the resource to get the params before making another resource */**
> ```
> 
> I beg your help to test it out fully and suggest changes or make new pull request to make it better.

I need something like this but for discount codes. right now the prev and next links are empty.

**Comment by damix1984** [comment]
Hi everyone,

I need to use pagination in a search call for GiftCards (/admin/gift_cards/search.json?query=updated_at:>"2019-01-01T00:00:00-05:00"), is it possible through this SDK?

Can you suggest me how?

I see the search method has only the query parameter... How do i pass to che method the pagination parameters? Am I missing something?
